app/Lib/Tools/EventTimelineTool.php

```diff
@@ -0,0 +1,138 @@
+        private $__related_events = array();
+        private $__related_attributes = array();
+
+        public function construct($eventModel, $user, $filterRules, $extended_view=0)
+        {
+            $this->__eventModel = $eventModel;
+            $this->__objectTemplateModel = $eventModel->Object->ObjectTemplate;
+            $this->__user = $user;
+            $this->__filterRules = $filterRules;
+            $this->__json = array();
+            $this->__extended_view = $extended_view;
+            $this->__lookupTables = array(
+                'analysisLevels' => $this->__eventModel->analysisLevels,
+                'distributionLevels' => $this->__eventModel->Attribute->distributionLevels
+            );
+            return true;
+        }
+
+        public function construct_for_ref($refModel, $user)
+        {
+            $this->__refModel = $refModel;
+            $this->__user = $user;
+            $this->__json = array();
+            return true;
+        }
+
+        private function __get_event($id)
+        {
+            $fullevent = $this->__eventModel->fetchEvent($this->__user, array('eventid' => $id, 'flatten' => 0, 'includeTagRelations' => 1, 'extended' => $this->__extended_view));
+            $event = array();
+            if (empty($fullevent)) {
+                return $event;
+            }
+
+            if (!empty($fullevent[0]['Object'])) {
+                $event['Object'] = $fullevent[0]['Object'];
+            } else {
+                $event['Object'] = array();
+            }
+
+            if (!empty($fullevent[0]['Attribute'])) {
+                $event['Attribute'] = $fullevent[0]['Attribute'];
+            } else {
+                $event['Attribute'] = array();
+            }
+
+            return $event;
+        }
+
+        public function get_timeline($id)
+        {
+            $event = $this->__get_event($id);
+            $this->__json['items'] = array();
+
+            if (empty($event)) {
+                return $this->__json;
+            }
+
+            if (!empty($event['Object'])) {
+                $object = $event['Object'];
+            } else {
+                $object = array();
+            }
+
+            if (!empty($event['Attribute'])) {
+                $attribute = $event['Attribute'];
+            } else {
+                $attribute = array();
+            }
+
+            // extract links and node type
+            foreach ($attribute as $attr) {
+                $toPush = array(
+                    'id' => $attr['id'],
+                    'uuid' => $attr['uuid'],
+                    'content' => $attr['value'],
+                    'event_id' => $attr['event_id'],
+                    'group' => 'attribute',
+                    'timestamp' => $attr['timestamp'],
+                    'first_seen' => $attr['first_seen'],
+                    'last_seen' => $attr['last_seen'],
+                );
+                array_push($this->__json['items'], $toPush);
```

Comment:
Why not just `$this->__json['items'][] = $toPush;`? It is even faster and I think also cleaner.

Reply:
Makes sense indeed.